File: csharp/direct-keys/DirectKeys.cs
/*
 * (c) 2016-2017 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://www.ionic.com/terms-of-use/)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

using System;
using System.Collections.Generic;

using IonicSecurity.SDK;

namespace DirectKeys
{
    class Program
    {
        static private Agent agent = null;

        static int Main(string[] args)
        {
            // Initialize the Ionic agent
            agent = new Agent();
            agent.Initialize();

            // Request keys
            // Forming the key request object
            CreateKeysRequest request = new CreateKeysRequest();
            // Here update request with the list of what it should create.
            AttributesDictionary dirAttributes = new AttributesDictionary();
            List<string> listClassValues = new List<string>(1);
            listClassValues.Add("restricted");
            dirAttributes.Add("classification", listClassValues);
            CreateKeysRequest.Key requestKey = new CreateKeysRequest.Key("reference_key", 2, dirAttributes);
            request.Keys.Add(requestKey);
            // Now ask the server to make those keys:
            CreateKeysResponse response;
            try
            {
                response = agent.CreateKeys(request);
            }
            catch (SdkException e)
            {
                System.Console.WriteLine("Error creating keys: {0}", e.Message);
                return -1;
            }

            // Show us what keys we got (you can always get a key right when you create it):
            List<CreateKeysResponse.Key> responseKeys = response.Keys;
            GetKeysRequest fetchRequest = new GetKeysRequest(); //we will use this to track the keys we want to fetch later
            foreach(CreateKeysResponse.Key responseKey in responseKeys)
            {
                System.Console.WriteLine("We created a key with the Key Tag: {0}", responseKey.Id);
                fetchRequest.KeyIds.Add(responseKey.Id);
            }

            // The rest of this program would typically happen at a different time,
            //  not right after creating the keys, but when you were going to access
            //  the data protected by those keys.

            // Now, using the Key Tags, ask the server for those keys again:
            // NOTE: We populated fetchRequest's list of keytags in the above loop.
            GetKeysResponse fetchResponse;
            try
            {
                fetchResponse = agent.GetKeys(fetchRequest);
            }
            catch (SdkException e)
            {
                System.Console.WriteLine("Error fetching keys: {0}", e.Message);
                return -1;
            }
            // Show what we got access to after a request for keys:
            foreach(GetKeysResponse.Key responseKey in fetchResponse.Keys)
            {
                System.Console.WriteLine("We fetched a key with the Key Tag: {0}", responseKey.Id);
            }

            // Tell us if we got less keys when we fetched than we created.
            //  This would happen if policy didn't give us access to all the keys.
            if (fetchResponse.Keys.Count < fetchRequest.KeyIds.Count)
            {
                System.Console.Write("We didn't get given all of the requested keys.");
                return -2;
            }

            System.Console.Read();
            return 0;
        }
    }
}

File: python/direct-keys/directkeys.py
#!/usr/bin/env python

# (c) 2017 Ionic Security Inc.
# By using this code, I agree to the Terms & Conditions (https://www.ionic.com/terms-of-use/)
# and the Privacy Policy (https://www.ionic.com/privacy-notice/).

import ionicsdk

import sys
import platform
import os.path


def get_agent():
  # Load a plaintext Device Profile, or SEP, from disk if on Linux (e.g., no default persistor).
  persistor = None
  if platform.system == 'Linux':
    SEP_FILE_PATH = os.path.expanduser("~/.ionicsecurity/profiles.pt")
    persistor = ionicsdk.DeviceProfilePersistorPlaintextFile(SEP_FILE_PATH)
    print("A plaintext SEP will be loaded from {0}".format(SEP_FILE_PATH))

  # Initialize the Ionic Agent (must be done before most Ionic operations).
  agent = ionicsdk.Agent(None, persistor)

  # Check if there are profiles and that there is one set as active.
  if not agent.hasanyprofiles() or not agent.hasactiveprofile():
    if not agent.hasanyprofiles():
      print("There are no device profiles on this device.")
    if not agent.hasactiveprofile():
      print("There is not an active device profile selected on this device.")
    print("Register (and select an active profile) this device before continuing.")
    sys.exit(-1)

  return agent


if __name__ == "__main__":
  agent = get_agent()
  if agent is None:
    sys.exit(-1)

  # Create a classification (or any other key attributes to set)
  key_attrs = {"classification": "Restricted"}
  # Add metadata to describe the decryption context
  req_metadata = {"example-metadata-key": "example-metadata-value"}

  # Now ask the server to make those keys:
  try:
    createdKeys = agent.createkeys(2, attributes=key_attrs, metadata=req_metadata)
  except ionicsdk.exceptions.IonicException as e:
    print("Error creating a key: {0}".format(e.message))
    sys.exit(-2)

  # Show us what keys we got (you can always get a key right when you create it):
  fetchRequest = [] # we will use this to track the keys we want to fetch later
  for key in createdKeys:
    # In Python, run `help(ionicsdk.agent.KeyData)` to learn about the key object.
    print("We created a key with the Key Tag: {0}".format(key.id))
    fetchRequest.append(key.id)

  # The rest of this program would typically happen at a different time,
  #  not right after creating the keys, but when you were going to access
  #  the data protected by those keys.

  # Now, using the Key Tags, ask the server for those keys again:
  # NOTE: We populated fetchRequest's list of keytags in the above loop.
  try:
    fetchedKeys = agent.getkeys(fetchRequest, metadata=req_metadata)
  except ionicsdk.exceptions.IonicException as e:
    print("Error fetching keys: {0}".format(e.message))
    sys.exit(-3)

  # Show what we got access to after a request for keys:
  for fetchedKey in fetchedKeys:
    print("We fetched a key with the Key Tag: {0}".format(fetchedKey.id))

  # Tell us if we got less keys when we fetched than we created.
  # This would happen if policy didn't give us access to all the keys.
  if len(fetchedKeys) != len(fetchRequest):
    print("We did not receive all of the requested keys.")
    sys.exit(-4)
